File: pinochlecore/src/main/scala/bboreel/games/card/game/Hand.scala
package bboreel.games.card.game

import bboreel.games.card.domain.Card

import scala.collection.mutable.ArrayBuffer

/**
  * Represents a players hand. Tracks the cards in the users hand and allow for dealing and
  * playing cards.
  * @author Robb Lee (robbmlee@gmail.com).
  */
class Hand extends {
  private var cards: ArrayBuffer[Card] = new ArrayBuffer[Card]()

  /**
    * Determines if the hand has any cards.
    *
    * @return [[Boolean]] true if the [[Hand]] has any cards else false.
    */
  def isEmpty(): Boolean = {
    cards.isEmpty
  }

  /**
    * Determines if the hand contains a card by definition.
    *
    * @param card [[Card]] object to check hand for.
    * @return [[Boolean]] true if the [[Card]] exists else false.
    */
  def contains(card: Card): Boolean = {
    cards.contains(card)
  }

  /**
    * Returns the number of cards in a hand.
    *
    * @return [[Int]] value of the number of cards in the hand.
    */
  def size(): Int = {
    cards.size
  }

  /**
    * Adds a card to the hand. Allows for adding multiple instances of the same card by reference
    * or as a seperate instance.
    *
    * @param card [[Card]] definition of a card to add to the hand.
    */
  def dealCard(card: Card): Unit = {
    cards.append(card)
  }

  /**
    * Plays a single card from the hand, if more than one card exists only a single card is played.
    * Note: This method does not protect against tying to play a non-existent card.
    *
    * @param card [[Card]] object to remove from the hand.
    */
  def playCard(card: Card): Card = {
    val cardIndex = cards.indexOf(card)
    if (cardIndex < 0) {
      throw new RuntimeException(s"Tried to remove invalid card: $card")
    }
    cards.remove(cardIndex)
  }
}

File: card-deck/src/main/java/org/eelbbor/carddeck/Hand.java
package org.eelbbor.carddeck;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Enables standard behaviors for a hand in a card game such as being dealt or playing a card.
 *
 * @author Robb Lee (robbmlee@gmail.com).
 */
public class Hand<T extends Card> {
  private List<T> cards;

  public Hand() {
    this.cards = new ArrayList<>();
  }

  /**
   * Indicates the number of cards in the hand.
   *
   * @return int value representing the number of cards in the hand.
   */
  public int numCards() {
    return cards.size();
  }

  /**
   * Indicates if a card is present in the hand.
   *
   * @param card {#link Card} to check the hand for.
   * @return true if the card is in the hand, else false.
   */
  public boolean containsCard(T card) {
    return cards.isEmpty() ? false : cards.contains(card);
  }

  /**
   * Adds a card or cards to the hand.
   *
   * @param card variable length number of cards to add to the hand.
   */
  public void dealCard(T... card) {
    cards.addAll(Arrays.asList(card));
  }

  /**
   * Removes the card if it is present in the hand. Unopinionated method does not throw
   * exception if
   * the card is not in the hand. Use the return value to indicate if the card was played.
   *
   * @param card {@link Card} object to play.
   * @return true if the card was in the hand, else false
   */
  public boolean playCard(T card) {
    return cards.remove(card);
  }

  /**
   * Returns a list of the cards remaining in the hand. Note: The return value is not tied to the
   * cards in the hands so will mutating it will have no side effect.
   *
   * @return list containing the remaining cards.
   */
  public List<T> remainingCards() {
    return new ArrayList<>(cards);
  }
}
